src/skydb.ts

```diff
@@ -271,6 +274,128 @@ export async function setDataLink(
+  // Rest of validation is done in `getEntry`.
+
+  const opts = {
+    ...defaultGetJSONOptions,
+    ...this.customOptions,
+    ...customOptions,
+  };
+
+  // Lookup the registry entry.
+  const entry = await getRegistryEntry(this, publicKey, dataKey, opts);
+  if (entry === null) {
+    return { data: null, dataLink: null };
+  }
+
+  // Determine the data link.
+  // TODO: Can this still be an entry link which hasn't yet resolved to a data link?
+  const { rawDataLink, dataLink } = parseDataLink(entry.data, false);
+
+  // If a cached data link is provided and the data link hasn't changed, return.
+  if (checkCachedDataLink(rawDataLink, opts.cachedDataLink)) {
+    return { data: null, dataLink };
+  }
+
+  // Download the data in the returned data link.
+  const downloadOpts = { ...extractOptions(opts, defaultDownloadOptions), responseType: "arraybuffer" as ResponseType };
+  const { data: buffer } = await this.getFileContent<ArrayBuffer>(dataLink, downloadOpts);
+
+  return { data: new Uint8Array(buffer), dataLink };
+}
+
+/* istanbul ignore next */
+/**
+ * Gets the registry entry for the given raw bytes or creates the entry if it doesn't exist.
+ *
+ * @param client - The Skynet client.
+ * @param publicKey - The user public key.
+ * @param dataKey - The dat akey.
+ * @param data - The raw byte data to set.
+ * @param [customOptions] - Additional settings that can optionally be set.
+ * @returns - The registry entry and corresponding data link.
+ * @throws - Will throw if the revision is already the maximum value.
+ */
+// TODO: Rename & refactor after the SkyDB caching refactor.
+export async function getOrCreateRawBytesRegistryEntry(
+  client: SkynetClient,
+  publicKey: string,
+  dataKey: string,
+  data: Uint8Array,
+  customOptions?: CustomSetJSONOptions
+): Promise<RegistryEntry> {
+  // Not publicly available, don't validate input.
+
+  const opts = {
+    ...defaultSetJSONOptions,
+    ...client.customOptions,
+    ...customOptions,
+  };
+
+  // Create the data to upload to acquire its skylink.
+  let dataKeyHex = dataKey;
+  if (!opts.hashedDataKeyHex) {
+    dataKeyHex = toHexString(stringToUint8ArrayUtf8(dataKey));
+  }
+  const file = new File([data], `dk:${dataKeyHex}`, { type: "application/octet-stream" });
+
+  // Start file upload, do not block.
+  const uploadOpts = extractOptions(opts, defaultUploadOptions);
+  const skyfilePromise: Promise<UploadRequestResponse> = client.uploadFile(file, uploadOpts);
+
+  // Fetch the current value to find out the revision.
+  //
+  // Start getEntry, do not block.
+  const getEntryOpts = extractOptions(opts, defaultGetEntryOptions);
+  const entryPromise: Promise<SignedRegistryEntry> = client.registry.getEntry(publicKey, dataKey, getEntryOpts);
+
+  // Block until both getEntry and uploadFile are finished.
+  const [signedEntry, skyfile] = await Promise.all<SignedRegistryEntry, UploadRequestResponse>([
```

Comment:
I'm worried about this: 

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAJwKZRMsAFZcC2MAzqgBREAWcIANgCYBKqRcNAbqgFyIBGcrqAIZgAlN1wFiqADxEoyGGADmAPkQBvAFCJEMYInJVajZqw4iN2nYiI1UqAA6kAjAAZXIgNxWAvlbQYWIhgqADuiBKEJKSkaCzsqAA0KKgAVqjQFgC8alrWuvqG1PRM8eaW+TpxZmQARAwAogDKAPIAMgBqDQAitV5W+QGYYN75foNpGVCk9Q0AUg0AwgAqPX2eiL79fpqgkLAINnaOpPhE3GAg+DyoyGIReFEybHAwdLkDQ0Eh4ZFSpCkyqhEDkbOhljB8KhqNNqglkmcLNtNJpSIIiABPSAGbIfHQQBDxVAAOhocCUM0WAEEmg1EM5uA5HlJiYIaDRECAiIolIhiVAKKgkMI6HyIIIoBAKH0rH8SKz2aQANpfHDM6LyECoETJVVysjANkkEQAXRE-MFYFIbDZILUBLARNJ5OtbJE5vFkoopFuyDtiAdTrJFNqvtqyV97qsVkEoUEMCgR3sTjc7hEKPxhIEzpD1NpiAATIz1SS2RzQgmKIhY-HE2j2YCaiIZZnHYnxSQC6VaFAiCDq3GEw9JPKy8q9SXSJrtbr0MN9VPkFrTenW0GXR3UF3mD2iOmrIHs8HKTS6QBmYsj0sNisCge1gwsKEpdKZFsBwnt9GoM-dmi9-sayHfUFRocc5ywBdDRoY1Z0CNUr0XZczQPLM7BzUhN1-Hd-z3RAAHp8MQZYAAkAEkmkQAA5BougYCIGDIqjliaTQfBEUgRCAA

Reply:
Yeah @peterjan doesn't that error just bubble up to the caller? Why would `case3Results` print in that scenario? What would be the benefit of catching that error here in `skydb.ts`? What is your proposal?

Reply:
I'm not expert on `Promise.all` but it seems that a `reject` is not throwing if it's not wrapped inside of a `try-catch` construct. Seeing as `getOrCreateRawBytesRegistryEntry` is in itself also not wrapped inside of a `try-catch`, and it sits right under `setJSONEncrypted`, I think there's a possibility that if a developer uses the SDK, without wrapping `setJSONEncrypted` in a try-catch himself, it will be in a deadlock if either of those promises fail..

The benefit of catching that error here would be so we can rethrow it and manage the behaviour from within the SDK, and not rely on a 3rd party developer to wrap an SDK call in a try-catch in order for it not to deadlock.

Again, I'm not sure if that concern is valid, but the worst-case is that this causes an application to just hang. So because that's such a terribly bad case, I think it warrants a closer look right? There's also something like `Promise.allSettled` which we might use here.

Another reason would be to avoid `Uncaught error in Promise` bubbling up from our SDK. I don't think we ever want to expose an uncaught promise rejection from our SDK. Instead you want to ensure all of those are catched, and then rethrow, possibly wrapping it in a more meaningful error.

Reply:
Thanks for the comment. After some discussion it looks like this doesn't warrant more attention.